### PR TITLE
Add QRCodeDetector to JavaScript Build

### PR DESCRIPTION
**Pull request by Ziachnix**
This PR adds javascript support for the class QRCodeDetector.
Tests for the proposed functions are derived from the equivalent python tests.
Because sample images are needed for running the tests, the build of the tests
requires now the opencv_extra repository.
(Alternatively i could copy some test images into the tutorial folder and use
them in the tests)
Due to emscripten limitations the functions _decodeMulti_ and
_detectAndDecodeMulti_ are excluded.
As far as i know it is not possible to use a function parameter (string array)
to return the decoded content of the QR codes.
 
### Pull Request Readiness Checklist
 
See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request
 
- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
build_image:Docs=docs-js
buildworker:Docs=linux-4,linux-6
build_image:Custom=javascript
buildworker:Custom=linux-4
```

### CODE REVIEW

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 1]
```diff
@@ -128,16 +128,20 @@ Building OpenCV.js from Source
 
     For example:
     @code{.bash}
-    emcmake python ./opencv/platforms/js/build_js.py build_js --build_test
+    emcmake python ./opencv/platforms/js/build_js.py build_js --build_test --cmake_option="-DOPENCV_TEST_DATA_PATH=./opencv_extra/testdata"
     @endcode
 
+    @note
+    It requires the [opencv_extra repository](https://github.com/opencv/opencv.git) located in your development environment.
```
Comment (alalek) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown:135]: Lets keep `OPENCV_TEST_DATA_PATH` optional.

Prefer to use environment variable for `OPENCV_TEST_DATA_PATH`.

File: modules/js/CMakeLists.txt [hunk 2]
```diff
@@ -110,15 +112,38 @@ foreach(f ${test_files})
...
+add_custom_command(OUTPUT "${opencv_test_js_bin_dir}/${test_data_haarcascade}"
+                   COMMAND ${CMAKE_COMMAND} -E copy_if_different "${test_data_haarcascade_path}"
+                   "${opencv_test_js_bin_dir}/${test_data_haarcascade}"
+                   DEPENDS "${test_data_haarcascade_path}"
+                   COMMENT "Copying ${test_data_haarcascade}"
+                  )
+list(APPEND opencv_test_js_file_deps "${test_data_haarcascade_path}" "${opencv_test_js_bin_dir}/${test_data_haarcascade}")
+# QR code
+set(test_data_qrcode_1 "link_ocv.jpg")
+set(test_data_qrcode_1_path "${OPENCV_TEST_DATA_PATH}/cv/qrcode/${test_data_qrcode_1}")
```
Comment (alalek) [inline modules/js/CMakeLists.txt:128]: Copy data only if `build_tests` enabled.

File: modules/js/generator/embindgen.py [hunk 3]
```diff
@@ -119,6 +119,7 @@ def makeWhiteList(module_list):
     'InputOutputArray': 'cv::Mat&',
     'InputArrayOfArrays': 'const std::vector<cv::Mat>&',
     'OutputArrayOfArrays': 'std::vector<cv::Mat>&',
+    'string':'std::string', # Side effect
     'String': 'std::string',
```
Comment (alalek) [inline modules/js/generator/embindgen.py:123]: Keep existed coding style.

Please remove the comment as it doesn't help and just confuse readers.

File: modules/js/test/test_objdetect.js [hunk 4]
```diff
@@ -72,11 +72,11 @@ if (typeof module !== 'undefined' && module.exports) {
     // The environment is Node.js
     var cv = require('./opencv.js'); // eslint-disable-line no-var
     cv.FS_createLazyFile('/', 'haarcascade_frontalface_default.xml', // eslint-disable-line new-cap
-                         'haarcascade_frontalface_default.xml', true, false);
+        'haarcascade_frontalface_default.xml', true, false);
```
Comment (alalek) [inline modules/js/test/test_objdetect.js:75]: Avoid unnecessary changes (here and below)

File: modules/js/test/test_objdetect.js [hunk 5]
```diff
@@ -159,3 +159,73 @@ QUnit.test('Cascade classification', function(assert) {
         locations.delete();
     }
 });
+QUnit.test('QR code detect and decode', function (assert) {
+    // QrCodeDetector
```
Comment (alalek) [inline modules/js/test/test_objdetect.js:163]: Comment looks useless.

File: modules/js/test/test_objdetect.js [hunk 6]
```diff
@@ -159,3 +159,73 @@ QUnit.test('Cascade classification', function(assert) {
...
+    {
+        assert.expect(12);
+        let done = assert.async(2);
+
+        const detector = new cv.QRCodeDetector();
+        const imgSingleQrCode = new Image();
+        imgSingleQrCode.crossOrigin = 'anonymous';
+        imgSingleQrCode.onload = function () {
+            const canvas = document.createElement('canvas');
+            const ctx = canvas.getContext('2d');
```
Comment (alalek) [inline modules/js/test/test_objdetect.js:173]: Does it work in Node.JS ?

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

Please take a look on comments below.

File: modules/js/generator/embindgen.py (hunk 3 above)
Comment (Ziachnix) [reply modules/js/generator/embindgen.py:123]: Thank you, i will remove the comment and fix the coding style.
If you swap the order of the entry `'string': 'std::string'` (line 122) with `'String': 'std::string'` (line 123) a double replacement happens and the code doesn't work anymore.
Should i add a comment there with a warning to not swap the order of the entries?

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown (hunk 1 above)
Comment (Ziachnix) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown:135]: Did i get you right, that the tests for the `QRCodeDetector` should be skipped when the enviroment variable `OPENCV_TEST_DATA_PATH` is not set? Because the tests depend on two files from the `opencv_extra` repository.

Comment (alalek) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown:135]: Right

File: modules/js/generator/embindgen.py (hunk 3 above)
Comment (alalek) [reply modules/js/generator/embindgen.py:123]: Thank you for information! I will take a look on this behavior

File: modules/js/test/test_objdetect.js (hunk 6 above)
Comment (Ziachnix) [reply modules/js/test/test_objdetect.js:173]: You are right, this is not working in Node.js.
In order to provide tests i have seen other JavaScript tests, where images are created from a hardcoded array.
In this test case it would probably be a huge array to create an image containing a QR-code.
Is there a way of loading images, which works both in the browser and in node.js? Otherwise i have to implement two different ways of loading the images.

**Comment by Ziachnix** [comment]
WIP: Sorry, something went wrong during the rebase

File: modules/js/test/test_objdetect.js (hunk 6 above)
Comment (alalek) [reply modules/js/test/test_objdetect.js:172]: Until we have good images (testdata) support in JS tests infrastructure, lets check the "bindings" existence only (without any attempt to perform accuracy checks).

Please use Mat.zeros(...) as input in tests for now. And check that there are no exceptions raised. Minimal smoke test.

For accuracy demo it would be nice to add tutorial page ([examples](https://docs.opencv.org/3.4/dc/d73/tutorial_js_table_of_contents_objdetect.html)).

**Comment by asmorkalov** [comment]
@Ziachnix Do you have progress with the PR?

File: modules/js/test/test_objdetect.js (hunk 6 above)
Comment (Ziachnix) [reply modules/js/test/test_objdetect.js]: I have now added tests that do not require any test images.
Therefore I have removed the copying of the test images again.
I will start writing a tutorial page. Should the tutorial include the QR code recognition algorithm? As far as I could understand `QrCodeDetector::detect` uses the algorithm described in this [article](https://www.atlantis-press.com/proceedings/emeit-12/3264).

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown (hunk 1 above)
Comment (alalek) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown:135]: Please drop `OPENCV_TEST_DATA_PATH` changes from the patch

**Review by alalek — COMMENTED** [review]
Thank you for update!

**Comment by asmorkalov** [comment]
@Ziachnix Friendly reminder.

**Review by alalek — APPROVED** [review]

**Comment by Ziachnix** [comment]
> 
> 
> @Ziachnix Friendly reminder.

Thank you, i totally forgot that there is still some work to be done.
I will open a new pull request when the tutorial page is ready to be merged.
